code/obj/item/pet_carrier.dm

```diff
@@ -0,0 +1,331 @@
+		var/carrier_occupants_length = length(src.carrier_occupants)
+		if (carrier_occupants_length)
+			if (carrier_occupants_length > src.explicit_name_limit)
+				. += "There's a whole zoo inside!"
+				return
+			. += "It's carrying [english_list(src.carrier_occupants)]."
+
+	update_icon()
+		..()
+		if (length(src.carrier_occupants))
+			src.grate_proxy.icon_state = src.grate_closed_icon_state
+			src.item_state = src.carrier_closed_item_state
+		else
+			src.grate_proxy.icon_state = src.grate_open_icon_state
+			src.item_state = src.carrier_open_item_state
+
+	attack(mob/M, mob/user)
+		if (user.a_intent == INTENT_HARM)
+			return ..()
+		if (istype(M))
+			if (!src.find_empty_hand(user))
+				boutput(user, "<span class='alert'>You need a free hand to scoop up [M]!</span>")
+				return ..()
+			if (!istype(M,src.allowed_mob_type))
+				boutput(user, "<span class='alert'>[M] can't quite fit inside [src]!</span>")
+				return ..()
+			if (src.carrier_max_capacity <= length(src.carrier_occupants))
+				boutput(user, "<span class='alert'>[src] is too crowded to fit one more!</span>")
+				return ..()
+			actions.start(new /datum/action/bar/icon/pet_carrier(M, src, src.icon, src.trap_mob_icon_state, TRAP_MOB, src.actionbar_duration), user)
+			return
+		..()
+
+	attack_self(mob/user)
+		src.attempt_removal(user)
+		..()
+
+	// Ensure that things inside can actually breathe.
+	remove_air(amount)
+		var/turf/current_turf = get_turf(src)
+		. = current_turf.remove_air(amount)
+
+	return_air()
+		var/turf/current_turf = get_turf(src)
+		. = current_turf.return_air()
+
+	mob_flip_inside(mob/user)
+		..(user)
+		src.mob_resist_inside(user)
+
+	mob_resist_inside(mob/user)
+		if (ON_COOLDOWN(src, "resist_damage", 3 SECONDS))
+			return
+		animate_storage_thump(src)
+		if (!src.can_break_out)
+			boutput(user, "<span class='alert'>It's no use! You can't leave [src]!</span>")
+			return
+		boutput(user, "<span class='alert'>You try to bust open the door of [src]!</span>")
+		src.take_door_damage(src.damage_per_resist)
+
+	throw_impact(atom/hit_atom, datum/thrown_thing/thr)
+		..()
+		if (src.carrier_occupants)
+			animate_storage_thump(src)
+		for (var/mob/occupant in src.carrier_occupants)
+			occupant.throw_impact(hit_atom, thr)
+		if (length(src.carrier_occupants))
+			src.take_door_damage(src.damage_per_resist * length(src.carrier_occupants))
+
+	/// Called when a given mob/user steals a mob after an actionbar.
+	proc/trap_mob(mob/mob_to_trap, mob/user)
+		if (!mob_to_trap)
+			return
+		if (mob_to_trap == user)
+			user.drop_item(src)
+		src.add_mob(mob_to_trap)
+		user.update_inhands()
+
+	/// Called when a given mob/user releases an mob after an actionbar.
+	proc/release_mob(mob/mob_to_release, mob/user)
+		if (mob_to_release)
+			src.eject_mob(mob_to_release)
+			user.update_inhands()
+			return
+		boutput(user, "<span class='alert'>Unable to release anyone from [src]!</span>")
+
+	proc/attempt_removal(mob/user)
+		if (!src.find_empty_hand(user))
```

Comment:
Does this mean that one armed people can never remove anything from the carrier?
If so, I'm fine removing it. I don't think people needing a free hand to do that is very relevant balance-wise.